packages/@aws-cdk/cdk/lib/cfn-reference.ts

```diff
@@ -0,0 +1,112 @@
+import { Reference } from "./reference";
+
+/**
+ * A Token that represents a CloudFormation reference to another resource
+ *
+ * If these references are used in a different stack from where they are
+ * defined, appropriate CloudFormation `Export`s and `Fn::ImportValue`s will be
+ * synthesized automatically instead of the regular CloudFormation references.
+ *
+ * Additionally, the dependency between the stacks will be recorded, and the toolkit
+ * will make sure to deploy producing stack before the consuming stack.
+ *
+ * This magic happens in the prepare() phase, where consuming stacks will call
+ * `consumeFromStack` on these Tokens and if they happen to be exported by a different
+ * Stack, we'll register the dependency.
+ */
+export class CfnReference extends Reference {
+  /**
+   * Check whether this is actually a Reference
+   */
+  public static isCfnReferenceToken(x: Token): x is CfnReference {
+    return (x as any).consumeFromStack !== undefined;
```

Comment:
I think you could instead use a private symbol to make this 100% reliable:

```typescript
// don't export
const isReference = Symbol.for('CfnReference');
export class CfnReference extends Reference {
  // private shouldn't bother JSII
  private [isReference] = true;

  public static isCfnReferenceToken(x: Token): x is CfnReference {
    return (x as any)[isReference] === true;
  }
}
```

Reply:
`private` doesn't bother JSII but it does bother TypeScript, which complains about unused private members.

Reply:
I am also not 100% sure that a private symbol will work across two different versions of this module (which is the reason we can't use `instanceof`).

My favorite (and very ugly still) way to do this is to add the marker at runtime:

```ts
const MARKER_KEY = '$type-marker';
const MARKER = '<uuid>';

export class MyClass {
  public static isMyClass(obj: any): obj is MyClass {
    return obj[MARKER_KEY] === MARKER;
  }

  constructor() {
    Object.defineProperty(this, MARKER_KEY, {
      value: MARKER,
      enumerable: false
    });
  }
}
```

Yeah, horrible. TypeScript, can you make this easier?

Reply:
It will work across two different versions because `Symbol.for('key')` stores the symbol in the global symbol repository by *name*. It would not work if you were to just call `Symbol()`, which is unique for every call. By private, I mean hiding it from JSII.

Reply:
We can use this to avoid some of the ugly duck typing we do - mark up the core constructs with symbols and maybe even generate some for L1. Module augmentation may even allow us to generate for our L2 later. 

Create a symbol for each type that would be checked by name in a nominal type system.

Reply:
Yes, `Symbol.for()` will fix it. Although I do wonder what the point of making it a Symbol is, then, since we lose the uniqueness value of it. I guess that it can't conflict with a naively string.

Reply:
On top of avoiding accidental conflicts, Symbols are not considered properties so it won't show up when enumerating an object's keys. Basically, it's cleaner and simpler.